primitives/state-machine/src/lib.rs

```diff
@@ -347,11 +347,11 @@ impl<'a, B, H, N, Exec> StateMachine<'a, B, H, N, Exec> where
 				CallResult<R, Exec::Error>,
 			) -> CallResult<R, Exec::Error>
 	{
-		let pending_changes = self.overlay.clone_pending();
+		self.overlay.start_transaction();
 		let (result, was_native) = self.execute_aux(true, native_call.take());
 
 		if was_native {
-			self.overlay.replace_pending(pending_changes);
+			self.overlay.rollback_transaction();
```

Comment:
I'm not familiar with this code, what is the reason for not committing whenever you return a result (like below)?

Reply:
Because in the case below we never started a transaction since we do not intend to roll it back. It is written directly to the parent transaction.